src/main/java/se/liu/ida/hefquin/engine/queryplan/utils/TextBasedPhysicalPlanPrinterImpl2.java

```diff
@@ -0,0 +1,293 @@
+		final LogicalOperator lop = op.getLogicalOperator();
+		printLogicalOperatorBase( "  - lop ", lop, out, indentLevelStringForOpDetail + singleBase );
+		out.append( System.lineSeparator() );
+		if ( lop instanceof LogicalOpTPAdd addOp ) {
+			printFederationMember( addOp.getFederationMember(), indentLevelStringForOpDetail + singleBase, out );
+		}
+		else if ( lop instanceof LogicalOpTPOptAdd addOp ) {
+			printFederationMember( addOp.getFederationMember(), indentLevelStringForOpDetail + singleBase, out );
+		}
+		printSPARQLGraphPattern( ((LogicalOpTPAdd) lop).getTP(), indentLevelStringForOpDetail + singleBase, out );
+		out.append( indentLevelStringForOpDetail + singleBase );
+		out.append( System.lineSeparator() );
+	}
+	
+	protected void printOperatorInfoForBindJoinWithFILTER ( final PhysicalOpBindJoinWithFILTER op, final PrintStream out, final String indentLevelString, final String indentLevelStringForOpDetail ) {
+		out.append( indentLevelString + "FILTERBindJoin (" + op.getID() + ") " );
+		out.append( System.lineSeparator() );
+		printLogicalOperatorBase( "  - lop ", op.getLogicalOperator(), out, indentLevelStringForOpDetail + singleBase );
+		out.append( System.lineSeparator() );
+	}
+	
+	protected void printOperatorInfoForBindJoinWithUNION ( final PhysicalOpBindJoinWithUNION op, final PrintStream out, final String indentLevelString, final String indentLevelStringForOpDetail ) {
+		out.append( indentLevelString + "UNIONBindJoin (" + op.getID() + ") " );
+		out.append( System.lineSeparator() );
+		final LogicalOperator lop = op.getLogicalOperator();
+		printLogicalOperatorBase( "  - lop ", lop, out, indentLevelStringForOpDetail + singleBase );
+		out.append( System.lineSeparator() );
+		printOperatorInfoFmAndPatternWithoutOpt( lop, out, indentLevelStringForOpDetail );
+	}
+	
+	protected void printOperatorInfoForBindJoinWithVALUES ( final PhysicalOpBindJoinWithVALUES op, final PrintStream out, final String indentLevelString, final String indentLevelStringForOpDetail ) {
+		out.append( indentLevelString + "VALUESBindJoin (" + op.getID() + ") " );
+		out.append( System.lineSeparator() );
+		final LogicalOperator lop = op.getLogicalOperator();
+		printLogicalOperatorBase( "  - lop ", lop, out, indentLevelStringForOpDetail + singleBase );
+		out.append( System.lineSeparator() );
+		printOperatorInfoFmAndPatternWithoutOpt( lop, out, indentLevelStringForOpDetail );
+	}
+	
+	protected void printOperatorInfoForOpFILTER ( final PhysicalOpFilter op, final PrintStream out, final String indentLevelString, final String indentLevelStringForOpDetail ) {
+		out.append( indentLevelString + "filter (" + op.getID() + ") " );
+		out.append( System.lineSeparator() );
+		final LogicalOperator lop = op.getLogicalOperator();
+		printLogicalOperatorBase( "  - lop ", lop, out, indentLevelStringForOpDetail + singleBase );
+		out.append( " filterExpressions (" + ((LogicalOpFilter) lop).getFilterExpressions().toString() + ")" );
```

Comment:
You can do the cast to `LogicalOpFilter` directly for `lop`. That makes hte code a bit more readable.
```suggestion
		final LogicalOpFilter lop = (LogicalOpFilter) op.getLogicalOperator();
		printLogicalOperatorBase( "  - lop ", lop, out, indentLevelStringForOpDetail + singleBase );
		out.append( " filterExpressions (" + lop.getFilterExpressions().toString() + ")" );
```

Reply:
... same in the cases of `LogicalOpGlobalToLocal` and `LogicalOpLocalToGlobal`.